packages/zip-it-and-ship-it/src/rate_limit.ts

```diff
@@ -1,30 +1,69 @@
-export enum RateLimitAlgorithm {
-  SlidingWindow = 'sliding_window',
-}
+import { z } from 'zod'
 
-export enum RateLimitAggregator {
-  Domain = 'domain',
-  IP = 'ip',
+export interface TrafficRules {
+  action: {
+    type: string
+    config: {
+      rateLimitConfig: {
+        algorithm: string
+        windowSize: number
+        windowLimit: number
+      }
+      aggregate: {
+        keys: {
+          type: string
+        }[]
+      }
+      to?: string
+    }
+  }
 }
 
-export enum RateLimitAction {
-  Limit = 'rate_limit',
-  Rewrite = 'rewrite',
-}
+const rateLimitAction = z.enum(['rate_limit', 'rewrite'])
+const rateLimitAlgorithm = z.enum(['sliding_window'])
+const rateLimitAggregator = z.enum(['domain', 'ip'])
+const slidingWindow = z.object({
+  windowLimit: z.number(),
+  windowSize: z.number(),
+})
+const rewriteActionConfig = z.object({
+  to: z.string(),
+})
 
-interface SlidingWindow {
-  windowLimit: number
-  windowSize: number
-}
+export const rateLimit = z
+  .object({
+    action: rateLimitAction.optional(),
+    aggregateBy: rateLimitAggregator.or(z.array(rateLimitAggregator)).optional(),
+    algorithm: rateLimitAlgorithm.optional(),
+  })
+  .merge(slidingWindow)
+  .merge(rewriteActionConfig.partial())
 
-export type RewriteActionConfig = SlidingWindow & {
-  to: string
-}
+type RateLimit = z.infer<typeof rateLimit>
 
-interface RateLimitConfig {
-  action?: RateLimitAction
-  aggregateBy?: RateLimitAggregator | RateLimitAggregator[]
-  algorithm?: RateLimitAlgorithm
-}
+/**
+ * Takes a rate limiting configuration object and returns a traffic rules
+ * object that is added to the manifest.
+ */
+export const getTrafficRulesConfig = (input: RateLimit): TrafficRules | undefined => {
+  const { windowSize, windowLimit, algorithm, aggregateBy, action, to } = input
+  const rateLimitAgg = Array.isArray(aggregateBy) ? aggregateBy : [rateLimitAggregator.Enum.domain]
```

Comment:
I think this incorrectly turns `rateLimitAggregator: "ip"` into `"domain"`. Should it be this instead?

```suggestion
  const rateLimitAgg = Array.isArray(aggregateBy) ? aggregateBy : [aggregateBy]
```

Reply:
I'm going by the current implementation: https://github.com/netlify/build/blob/461e1cad8ad1362901c3049a431281825bac9380/packages/zip-it-and-ship-it/src/runtimes/node/in_source_config/index.ts#L113

Reply:
Alright, then let's keep the fixing of that to a different PR. @paulo I think you wrote this (and I reviewed ...), what are your thoughts on it?

Reply:
The thought process at the time was to not fail the build and just use the default since it's not an obligatory field, but I'm now regretting it. If I could do it all over again I think I'd just fail the build if aggregateBy is not an array (if defined, not an obligatory field), but leaving the current logic is also ok